lib/browser/chrome-devtools.js

```diff
@@ -0,0 +1,121 @@
+'use strict'
+
+const { dialog, Menu } = require('electron')
+const fs = require('fs')
+const url = require('url')
+
+const ipcMainUtils = require('@electron/internal/browser/ipc-main-internal-utils')
+
+const convertToMenuTemplate = function (event, items) {
+  return items.map(function (item) {
+    const transformed = item.type === 'subMenu' ? {
+      type: 'submenu',
+      label: item.label,
+      enabled: item.enabled,
+      submenu: convertToMenuTemplate(event, item.subItems)
+    } : item.type === 'separator' ? {
+      type: 'separator'
+    } : item.type === 'checkbox' ? {
+      type: 'checkbox',
+      label: item.label,
+      enabled: item.enabled,
+      checked: item.checked
+    } : {
+      type: 'normal',
+      label: item.label,
+      enabled: item.enabled
+    }
+
+    if (item.id != null) {
+      transformed.click = function () {
+        event._replyInternal('ELECTRON_INSPECTOR_CONTEXT_MENU_CLICK', item.id)
+      }
+    }
+
+    return transformed
+  })
+}
+
+const getEditMenuItems = function () {
+  return [
+    { role: 'undo' },
+    { role: 'redo' },
+    { type: 'separator' },
+    { role: 'cut' },
+    { role: 'copy' },
+    { role: 'paste' },
+    { role: 'pasteAndMatchStyle' },
+    { role: 'delete' },
+    { role: 'selectAll' }
+  ]
+}
+
+const isChromeDevTools = function (pageURL) {
+  const { protocol } = url.parse(pageURL)
+  return protocol === 'chrome-devtools:'
+}
+
+const assertChromeDevTools = function (contents, api) {
+  const pageURL = contents._getURL()
+  if (!isChromeDevTools(pageURL)) {
+    console.error(`Blocked ${pageURL} from calling ${api}`)
+    throw new Error(`Blocked ${api}`)
+  }
+}
+
+ipcMainUtils.handle('ELECTRON_INSPECTOR_CONTEXT_MENU', function (event, items, isEditMenu) {
+  assertChromeDevTools(event.sender, 'window.InspectorFrontendHost.showContextMenuAtPoint()')
+
+  const template = isEditMenu ? getEditMenuItems() : convertToMenuTemplate(event, items)
+  const menu = Menu.buildFromTemplate(template)
+  const window = event.sender.getOwnerBrowserWindow()
+
+  menu.once('menu-will-close', () => {
+    setTimeout(() => {
```

Comment:
Why does this need a `setTimeout`? Can we reply immediately on close?

Reply:
@nornagon the close event comes before the click handler, it needs to be delayed

Reply:
does `setTimeout` guarantee that ordering, or just make it more likely?

what happens if they're misordered?

if this really is the right way to do it, please add a comment about why, otherwise this will be lost when this code next gets refactored.

Reply:
@nornagon I will add the comment. I don’t like the timeout, ideally we would have an event like `dismissed` indicating that no item was selected. The original version of the code aso had a bug as it didn’t let the devtools know when the menu was closed. So if you would right click several times, dismiss the menu and last time click on an item, it would trigger the selected action multiple times